torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp

```diff
@@ -567,256 +370,390 @@ Value* findArgumentAsInputParam(
+void InplaceConverter::gatherAttrNameInitialValueMap(
     Block* block,
-    const std::shared_ptr<Graph>& graph,
-    std::unordered_map<std::string, Value*>& allAttrValues,
-    std::unordered_map<std::string, Value*>& setAttrValues,
-    MutationRemover& mr,
-    Module* module_ = nullptr) {
-  Node* m = *block->nodes().begin();
-  WithInsertPoint guard(m);
-  std::unordered_map<std::string, Value*> nextSetAttrValues = {};
+    std::unordered_map<std::string, Value*>& attr_name_value_map,
+    std::unordered_map<Node*, std::string>& attr_node_fullname_map) {
+  for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+    Node* n = *it;
+    it++; // node n can be destroyed
+
+    for (auto* sub_block : n->blocks()) {
+      gatherAttrNameInitialValueMap(
+          sub_block, attr_name_value_map, attr_node_fullname_map);
+    }
+
+    if (n->kind() != prim::GetAttr && n->kind() != prim::SetAttr)
+      continue;
 
+    auto name = n->s(attr::name);
+    auto attrModule = *module_;
+    Value* paramConst = nullptr;
+
+    auto moduleNames =
+        findSubModuleAttr(n->inputs().at(0), name, attrModule, graph_);
+
+    std::string fullName("");
+    for (auto& name : moduleNames) {
+      fullName += name + '.';
+    }
+    fullName += name;
+
+    attr_node_fullname_map.insert({n, fullName});
+
+    if (attr_name_value_map.find(fullName) == attr_name_value_map.end() &&
+        attrModule.hasattr(name)) {
+      auto attr = attrModule.attr(name);
+      auto type = attrModule.type();
+      auto slot = *type->findAttributeSlot(name);
+
+      // Add model_parameters and model_buffers as model inputs. Order is
+      // preserved based on the appearance in the graph.
+      WithInsertPoint guard(graph_->nodes().front());
+      if (type->is_parameter(slot) || type->is_buffer(slot) ||
+          (attr.isObject() && !attr.toObjectRef().type()->is_module())) {
+        paramConst = findArgumentAsInputParam(graph_, fullName, attr);
+        attr_name_value_map.insert({fullName, paramConst});
+      } else if (auto attrVal = tryInsertConstant(*graph_, attr)) {
+        for (size_t i = 0; i < type->getAttributes().size(); i++) {
+          if (type->getAttributeName(i) == name) {
+            paramConst = *attrVal;
+            attr_name_value_map.insert({fullName, paramConst});
+          }
+        }
+      } else {
+        GRAPH_DEBUG(
+            attr.type()->cast<ClassType>() ? "" : "attribute: ",
+            name,
+            " is not materializable.");
+      }
+    }
+
+    // Create dummy initial value.
+    if (attr_name_value_map.find(fullName) == attr_name_value_map.end()) {
+      auto* noneNode = graph_->create(prim::Constant);
+      noneNode->output()->setType(NoneType::get());
+      noneNode->insertBefore(graph_->nodes().front());
+      attr_name_value_map.insert({fullName, noneNode->output()});
+    }
+  }
+}
+
+// Replace prim::GetAttr and prim::SetAttr with ATen inplace operators.
```

Comment:
Can you add a graph example here in comments?

Reply:
added